Run each makeflow rule as an Umbrella job

hmeng-19 opened the pull request:
Add `--umbrella-spec` and `--umbrella-binary` option into makeflow
Add `dag_node->umbrella_spec` to track the umbrella spec needed for each rule
Add `makeflow_wrapper_umbrella` model

hmeng-19 commented:
@btovar , @nhazekam , @dthain , PTAL.

nhazekam commented on dttools/src/path.h:
```diff
@@ -44,4 +44,9 @@ Returns NULL if none is found.
+/* path_join_two_strings joins two strings, and adds sep between them.
+ * The caller should free the space pointed by the returned pointer.
+ */
+char *path_join_two_strings(const char *s, const char *t, const char * sep);
```
Could also be useful for writing remote rename.

nhazekam commented on makeflow/src/makeflow.c:
```diff
@@ -1599,6 +1626,19 @@ int main(int argc, char *argv[])
 		makeflow_gc_method = MAKEFLOW_GC_ALL;
 	}
 
+	if(wrapper_umbrella && wrapper_umbrella->spec) {
```
For now umbrella should also require that the binary is specified. I understand that there may be cases when umbrella exists locally, but in Makeflow we assume that you are specifying the executable files in most cases. This also ensures consistent version across nodes.

hmeng-19 replied:
@nhazekam , the logic here is that if the user does not set up the `--umbrella-spec`, it is not necessary to set up the umbrella wrapper.
That means, if the user only sets `--umbrella-binary` without setting `--umbrella-spec`, we do nothing.

nhazekam replied:
You also never require the binary to be set. I understand that we only want to do something if spec is set, but for Makeflow we should also have the binary. There is a comment on this in the umbrella wrapper. Assuming we need both, then the logic should check for both.

hmeng-19 replied:
This is deliberate. So that the code is not broken no matter the umbrella binary is set or not.
If the umbrella binary is set, we will send the binary to the execution node and the execution node will run `./umbrella run ...`;
If the umbrella binary is not set, we will not send the binary to the execution node and the execution node will run `umbrella run ...`.

nhazekam replied:
After further discussion, not specifying the executable is fine, but if it is not specified we should emit a warning that umbrella will be used at the execution site.

nhazekam commented on makeflow/src/makeflow.c:
```diff
@@ -1599,6 +1626,19 @@ int main(int argc, char *argv[])
+		struct dag_node *cur;
+		cur = d->nodes;
+		while(cur) {
+			dag_node_set_umbrella_spec(cur, wrapper_umbrella->spec);
```
This seems redundant. I understand having an umbrella spec stored in the node for the eventuality that each node could have a different spec, but when they all share the same umbrella spec it seems to make more sense to only specify it in the umbrella wrapper. Then in the future when node can have different umbrella specs, the dag_node spec is only set when it is different from the standard.

hmeng-19 replied:
@nhazekam , I got your point.
The only reason I keep it is it does not really break anything. And I indeed have plan to adding support for specifying umbrella specs for each rule.
I would keep it for now if redundancy is your only concern here.

nhazekam replied:
If we are ever setting for each specific node then there will be no need to loop over them all and set the same one (we could still use the generic one in the wrapper). This is duplicate information and it isn't used in the current implementation.

hmeng-19 replied:
Okay. If you really want, I can remove it for now. 
But the next PR will definitely have these code again.

nhazekam commented on makeflow/src/makeflow_wrapper_umbrella.c:
```diff
@@ -0,0 +1,169 @@
+
+void makeflow_wrapper_umbrella_set_spec(struct makeflow_wrapper_umbrella *w, const char *spec) {
+	struct stat st;
+	if(lstat(spec, &st) == -1) {
```
Currently in Makeflow we use batch_fs_stat to check consistently with the batch system. Check makeflow.c for two examples.

hmeng-19 replied:
@nhazekam , the prototype of `batch_fs_stat` requires a `struct batch_queue *`:
```
int batch_fs_stat (struct batch_queue *q, const char *path, struct stat *buf)
```

`makeflow_wrapper_umbrella_set_spec` is called during the step of parsing options, when the batch_queue has not been initialized yet.

nhazekam commented on makeflow/src/makeflow_wrapper_umbrella.c:
```diff
@@ -0,0 +1,169 @@
+	if(lstat(spec, &st) == -1) {
+		fatal("lstat(`%s`) failed: %s\n", spec, strerror(errno));
+	}
+	if((st.st_mode & S_IFMT) != S_IFREG) {
```
@dthain @btovar Should we start checking if the file is regular or is this overkill for Makeflow?

nhazekam commented on makeflow/src/makeflow_wrapper_umbrella.c:
```diff
@@ -0,0 +1,169 @@
+
+void makeflow_wrapper_umbrella_set_binary(struct makeflow_wrapper_umbrella *w, const char *binary) {
+	struct stat st;
+	if(lstat(binary, &st) == -1) {
```
Same as above.

nhazekam commented on makeflow/src/makeflow_wrapper_umbrella.c:
```diff
@@ -0,0 +1,169 @@
+
+	// construct umbrella_command
+	if(!batch_queue_supports_feature(queue, "remote_rename")) {
+		if(!w->binary) {
```
Same as above, for consistency we should require that the binary is specified. This can be discussed, but in general with Makeflow we don't assume information about the execution environment. This is different when using programs like Docker that need special permissions to be installed, but when its software we are writing I think we can require it to be specified. @btovar @dthain

hmeng-19 replied:
My point for this is:
If the user provides the umbrella binary via `--umbrella-binary`, then we use it.
If the user does not provide the umbrella binary via `--umbrella-binary`, we expect to find the umbrella binary through `$PATH` on the execution node. This is useful when we later run work_queue_worker as an umbrella job, which delivers both work_queue_worker and umbrella .

nhazekam replied:
I understand this is a we previously discussed. However, I did comment then that in Makeflow we typically specify the executable. When assuming there is an executable for umbrella at the worker we are also assuming the version is correct and consistent with all other nodes.

nhazekam replied:
As with the above comment, this is allowable with the knowledge that the user should be warned that we are assuming the executable exists at the work site. Disregard my previous comments on this.

nhazekam commented:
As a general comment, how much has this been tested? Also has it been tested with other wrappers to see how they interact and that wrappers still work in the absence of umbrella. The last one should be fine, but we should still test it.

hmeng-19 commented:
@nhazekam , here are the testings for this PR, which is under `cctools/makeflow/example/makeflow_umbrella`:
```
To test makeflow with umbrella using local execution engine:
$ makeflow --umbrella-binary $(which umbrella) --umbrella-spec convert_S.umbrella example.makeflow

To test makeflow with umbrella using wq execution engine:
$ makeflow -T wq --umbrella-binary $(which umbrella) --umbrella-spec convert_S.umbrella example.makeflow

To test the case when `--wrapper` and `--umbrella-spec|binary` are used at the same time:
$ makeflow --wrapper 'time -p /bin/sh -c []'  --umbrella-binary $(which umbrella) --umbrella-spec convert_S.umbrella example.makeflow
$ makeflow -T wq --wrapper 'time -p /bin/sh -c []'  --umbrella-binary $(which umbrella) --umbrella-spec convert_S.umbrella example.makeflow
```
For each case, I also tested the case when the `--umbrella-binary` option is not set.

hmeng-19 commented:
@nhazekam , https://github.com/cooperative-computing-lab/cctools/pull/1458/commits/6798f4e9e0a7bdae50b11f3ef9081145a6ea896e  and https://github.com/cooperative-computing-lab/cctools/pull/1458/commits/8fdeb0a36f26bc57dd65f9528ebdddfd77cea9b7 should fix your concerns, PTAL.

hmeng-19 commented:
@nhazekam , fixed the conflicts and squashed the commits into a single one, as show in https://github.com/cooperative-computing-lab/cctools/pull/1458/commits/1866d35bea34f0281a5a1bceabacd5d409a7f7cd. PTAL.

nhazekam commented:
When testing it everything worked. The only issue is that when I run makeflow -c example.makeflow, the umbrella.log is left behind and not cleaned. This file should be added to the log so it is properly removed.

nhazekam commented:
We can merge. I'll open an issue about the umbrella.log.
